### PR TITLE
qscintilla2 with qt5 support

### PR DESCRIPTION
**Pull request by schoeps**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally prior to submission with `brew install <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [x] Does your submission pass `brew audit --strict --online <formula>` (after doing `brew install <formula>`)?

---

This pull request addresses https://github.com/Homebrew/legacy-homebrew/issues/37591 and https://github.com/Homebrew/legacy-homebrew/issues/41464. This changset als sets python as the default switch for pyqt5 (i.e., not python3) since this seems more consistent with all other formulae.

### CODE REVIEW

File: Formula/pyqt5.rb [hunk 1]
```diff
@@ -3,6 +3,7 @@ class Pyqt5 < Formula
   homepage "https://www.riverbankcomputing.com/software/pyqt/download5"
   url "https://downloads.sourceforge.net/project/pyqt/PyQt5/PyQt-5.6/PyQt5_gpl-5.6.tar.gz"
   sha256 "2e481a6c4c41b96ed3b33449e5f9599987c63a5c8db93313bd57a6acbf20f0e1"
+  revision 2
```
Comment (MikeMcQuaid) [inline Formula/pyqt5.rb]: Should be `1`

**Comment by MikeMcQuaid** [comment]
Installation is failing on all platforms: https://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/6046/version=el_capitan/testReport/junit/brew-test-bot/el_capitan/install_qscintilla2/



> Have you built your formula locally prior to submission with brew install <formula> (where <formula> is the name of the formula you're submitting)?

Did this work for you?

**Comment by schoeps** [comment]
Yes it did, that's why I made a cross in the box...

I looked at the results. Error is the following
```
mkdir: /usr/local/Cellar/qt/4.8.7_2/qsci/api/python/: Operation not permitted
mkdir: /usr/local/opt/pyqt/share/sip/Qsci/: Operation not permitted
```
There seems to be an issue  with the installation paths of qscintilla. My local installation did not care about this.

**Comment by MikeMcQuaid** [comment]
You can replicate this behaviour with `export HOMEBREW_SANDBOX=1`; it's a security feature that we need fixed before we can merge.

**Comment by schoeps** [comment]
is the sandbox rule new? I believe the original qscintilla2 was already broken.

**Comment by schoeps** [comment]
...and I am afraid I am not able to clean this up.

**Comment by schoeps** [comment]
I am surprised but if it's ok for Jenkins then it's ok for me :)

File: Formula/pyqt5.rb [hunk 2]
```diff
@@ -15,8 +16,8 @@ class Pyqt5 < Formula
 
   deprecated_option "enable-debug" => "with-debug"
 
-  depends_on :python3 => :recommended
-  depends_on :python => :optional
+  depends_on :python => :recommended
+  depends_on :python3 => :optional
```
Comment (MikeMcQuaid) [inline Formula/pyqt5.rb]: I remembered why they were the way round they were: otherwise the bottles will always require installing Homebrew's Python.

Comment (schoeps) [inline Formula/pyqt5.rb]: Now the bottle is always asking for Homebrew's python3, right? This is not too great, either?

I could do it as in https://github.com/Homebrew/homebrew-core/blob/master/Formula/pyqt.rb:
```
option "without-python", "Build without python 2 support"
depends_on :python3 => :optional
```
but is the loop `Language::Python.each_python(build) do |python, version|` then still taking care of python 2?

Comment (MikeMcQuaid) [inline Formula/pyqt5.rb]: We pretty much always mandate our own Python 3 so we've not had complaints about that.

I think it's worth just leaving this file as-is for now (or at least pulling into another PR)

Comment (schoeps) [inline Formula/pyqt5.rb]: I dropped my changeset to pyqt5. However, things are still inconsistent among pyqt,pyqt5 and qscintilla2.

Therefore I still ask in `qscintilla2.rb` for explicit `pyqt --with-python3` and `pyqt5 --with-python3` (albeit the latter is default, the former not).

File: Formula/qscintilla2.rb [hunk 3]
```diff
@@ -11,26 +12,30 @@ class Qscintilla2 < Formula
...
-    depends_on "pyqt" => "with-python3"
-  elsif build.with? "python"
-    depends_on "pyqt"
+  if build.with?("python") || build.with?("python3")
+    depends_on "sip" => ("with-python3" if build.with?("python3"))
+    depends_on "pyqt5" => ("with-python3" if build.with?("python3")) if build.with?("qt5")
+    depends_on "pyqt" => ("with-python3" if build.with?("python3")) if build.without?("qt5")
   else
-    depends_on "qt"
+    depends_on "qt5" if build.with?("qt5")
```
Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: `depends_on "qt5" => :optional`

File: Formula/qscintilla2.rb [hunk 4]
```diff
@@ -11,26 +12,30 @@ class Qscintilla2 < Formula
...
+    depends_on "qt" if build.without?("qt5")
```
Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: `depends_on "qt" => :recommended`

File: Formula/qscintilla2.rb [hunk 5]
```diff
@@ -11,26 +12,30 @@ class Qscintilla2 < Formula
...
   option "without-python", "Skip building the Python bindings"
 
   depends_on :python => :recommended
   depends_on :python3 => :optional
 
-  if build.with? "python3"
-    depends_on "pyqt" => "with-python3"
-  elsif build.with? "python"
-    depends_on "pyqt"
+  if build.with?("python") || build.with?("python3")
```
Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: Can you split the `build.with?("python3")` into another `if`; it's a bit hard to follow this as-is.

**Comment by MikeMcQuaid** [comment]
We're dropping Qt4 support very soon so another option to simplify this would be to always build against Qt5.

File: Formula/qscintilla2.rb [hunk 6]
```diff
@@ -59,25 +64,33 @@ def install
     if build.with?("python") || build.with?("python3")
       cd "Python" do
         Language::Python.each_python(build) do |python, version|
+          pyargs = ["--apidir=#{prefix}/qsci"]
+          pyargs << "--spec=#{spec}"
+          pyargs << "--qsci-sipdir=#{share}/sip"
+          if build.with?("qt5")
+            pyargs << "--pyqt=PyQt5"
+            pyargs << "--destdir=#{lib}/python#{version}/site-packages/PyQt5"
+            pyargs << "--pyqt-sipdir=#{Formula["pyqt5"].opt_prefix}/share/sip/Qt5"
```
Comment (DomT4) [inline Formula/qscintilla2.rb]: You're looking for `opt_share` here rather than `opt_prefix/share`. Is it actually installing anything there or is this purely a reference?

File: Formula/qscintilla2.rb [hunk 7]
```diff
@@ -59,25 +64,33 @@ def install
...
+          else
+            pyargs << "--pyqt=PyQt4"
+            pyargs << "--destdir=#{lib}/python#{version}/site-packages/PyQt4"
+            pyargs << "--pyqt-sipdir=#{Formula["pyqt"].opt_prefix}/share/sip"
```
Comment (DomT4) [inline Formula/qscintilla2.rb]: Same thing re `opt_share`

File: Formula/qscintilla2.rb [hunk 8]
```diff
@@ -88,10 +101,17 @@ def install
   end
 
   test do
-    Pathname("test.py").write <<-EOS.undent
-      import PyQt4.Qsci
-      assert("QsciLexer" in dir(PyQt4.Qsci))
-    EOS
+    if build.with?("qt5")
+      Pathname("test.py").write <<-EOS.undent
```
Comment (DomT4) [inline Formula/qscintilla2.rb]: Don't need the `Pathname`. Just `(testpath/"test.py").write <<-EOS.undent` etc.

**Comment by schoeps** [comment]
some recommendation seem to have broken qt5 :(

**Comment by schoeps** [comment]
ok, the problem is that `qscintilla2 --with-python3` also tries to compile for python2 which is not necessarily included in pyqt5. What to do? I would need to distinguish four cases: `--with-python`, `--with-python3`, `--with-python` & `--with-python3`, `--without-python`....

File: Formula/qscintilla2.rb [hunk 9]
```diff
@@ -11,26 +12,38 @@ class Qscintilla2 < Formula
...
   depends_on :python => :recommended
   depends_on :python3 => :optional
 
-  if build.with? "python3"
-    depends_on "pyqt" => "with-python3"
-  elsif build.with? "python"
-    depends_on "pyqt"
+  if build.with?("python") && build.with?("python3")
+    depends_on "sip" => "with-python3" # python2 is recommended
+    depends_on "pyqt5" => "with-python" if build.with?("qt5") # python3 is recommended
```
Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]:
```ruby
if build.with?("qt5")
  depends_on "pyqt5" => "with-python"
else
  depends_on "pyqt" => "with-python3"
end
```

Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: (and same below)

File: Formula/qscintilla2.rb [hunk 10]
```diff
@@ -11,26 +12,38 @@ class Qscintilla2 < Formula
...
 
   def install
-    # On Mavericks we want to target libc++, this requires an
-    # unsupported/macx-clang-libc++ flag.
     if ENV.compiler == :clang && MacOS.version >= :mavericks
-      spec = "unsupported/macx-clang-libc++"
-    else
+      if build.with?("qt5")
+        spec = "macx-clang"
+      else # if Qt4
```
Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: Can drop this comment.

File: Formula/qscintilla2.rb [hunk 11]
```diff
@@ -11,26 +12,38 @@ class Qscintilla2 < Formula
...
+        spec = "unsupported/macx-clang-libc++"
+      end
+    else # if gcc and (Qt4 or Qt5)
```
Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: Can drop this comment.

File: Formula/qscintilla2.rb [hunk 12]
```diff
@@ -11,26 +12,38 @@ class Qscintilla2 < Formula
...
   else
-    depends_on "qt"
+    depends_on "qt5" => :optional
+    depends_on "qt" => :recommended
   end
 
   def install
-    # On Mavericks we want to target libc++, this requires an
-    # unsupported/macx-clang-libc++ flag.
     if ENV.compiler == :clang && MacOS.version >= :mavericks
```
Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: Could do `spec = if ENV.compiler == :clang && MacOS.version >= :mavericks` and drop the `spec =` below.

File: Formula/qscintilla2.rb [hunk 13]
```diff
@@ -59,25 +72,34 @@ def install
...
+          system python, "configure.py", "-o", lib, "-n", include, *pyargs
           system "make"
           system "make", "install"
           system "make", "clean"
         end
       end
     end
 
-    if build.with? "plugin"
+    if build.with?("plugin")
```
Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: No need to change stuff like this.

**Comment by MikeMcQuaid** [comment]
Needs rebased on `master` as well as my tweaks. Cases look good so far 👍.

File: Formula/qscintilla2.rb (hunk 12 above)
Comment (schoeps) [inline Formula/qscintilla2.rb]: Sorry, I don't get this one.

Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: Instead if `if foo` with a `spec = ` below` you can do `spec if foo` in Ruby and rely on the implicit return from statements.

**Comment by schoeps** [comment]
@MikeMcQuaid: I think everything is rather nice now but do you happen to know why python does not find Qsci anymore?
```
==> python test.py
Traceback (most recent call last):
  File "test.py", line 1, in <module>
    import PyQt5.Qsci
ImportError: No module named PyQt5.Qsci
Error: qscintilla2: failed
Failed executing: python test.py
```

File: Formula/qscintilla2.rb [hunk 14]
```diff
@@ -42,11 +42,14 @@ def install
...
         s.gsub! "$$[QT_INSTALL_TRANSLATIONS]", prefix/"trans"
         s.gsub! "$$[QT_INSTALL_DATA]", prefix/"data"
+        s.gsub! "$$[QT_HOST_DATA]", prefix/"data"
+        #s.gsub! "TARGET = qscintilla2", "TARGET = qscintilla2-qt5"
       end
 
       inreplace "features/qscintilla2.prf" do |s|
         s.gsub! "$$[QT_INSTALL_LIBS]", lib
         s.gsub! "$$[QT_INSTALL_HEADERS]", include
+        #s.gsub! "LIBS += -lqscintilla2", "LIBS += -lqscintilla2-qt5"
```
Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: Can remove this commented-out line.

File: Formula/qscintilla2.rb [hunk 15]
```diff
@@ -42,11 +42,14 @@ def install
         s.gsub! "$$[QT_INSTALL_HEADERS]", include
         s.gsub! "$$[QT_INSTALL_TRANSLATIONS]", prefix/"trans"
         s.gsub! "$$[QT_INSTALL_DATA]", prefix/"data"
+        s.gsub! "$$[QT_HOST_DATA]", prefix/"data"
+        #s.gsub! "TARGET = qscintilla2", "TARGET = qscintilla2-qt5"
```
Comment (MikeMcQuaid) [inline Formula/qscintilla2.rb]: Can remove this commented-out line.

**Comment by MikeMcQuaid** [comment]
Tests are 💚. A couple of final nits then happy to 🚢 this.

**Comment by MikeMcQuaid** [comment]
Thanks @schoeps!

**Comment by DomT4** [comment]
Was this tested against `sqliteman` & `codequery` to ensure having them linked recursively to two different Qt branches didn't cause anything odd? The latter should be using `qt5` already, but IIRC development on the former is pretty 💀.

**Comment by MikeMcQuaid** [comment]
@DomT4 It was not. At this point I figured something without broken CI that crops up later was going to become either fixed or boneyarded shortly anyway.

**Comment by ruben2020** [comment]
Shouldn't you have another version of qscintilla2 for qt4? You can call it `qscintilla2-qt4`

**Comment by ilovezfs** [comment]
@ruben2020 it's pretty unlikely we'll ship something like that even in homebrew/versions. Is there a reverse dependency you have in mind that you think requires qscintilla2 and qt4?

**Comment by ruben2020** [comment]
I'm the owner of `codequery`. But it's no problem for me to switch to qt5. I've been building it for qt5 experimentally before anyway.

**Comment by ilovezfs** [comment]
@ruben2020 Cool! I assume you've seen https://github.com/Homebrew/homebrew-core/pull/4258

**Comment by ruben2020** [comment]
Thanks @ilovezfs . I didn't know about it. I will check it out.
